contracts/RewardPool.sol

```diff
@@ -92,65 +93,112 @@ contract RewardPool is Initializable, Versionable, Ownable {
     return true;
   }
 
+  function valid(
+    bytes memory leaf,
+    bytes32[] memory proof
+  ) public view returns (bool) {
+        (
+      address rewardProgramID,           // solhint-disable-line no-unused-vars
+      uint256 paymentCycleNumber,
+      uint256 startBlock,                // solhint-disable-line no-unused-vars
+      uint256 endBlock,                  // solhint-disable-line no-unused-vars
+      uint256 tokenType,                 // solhint-disable-line no-unused-vars
+      address payee,                     // solhint-disable-line no-unused-vars
+      bytes memory transferDetails       // solhint-disable-line no-unused-vars
+    ) = abi.decode(leaf, (address, uint256, uint256, uint256, uint256, address, bytes));
+    bytes32 root = bytes32(payeeRoots[paymentCycleNumber]);
+    return proof.verify(root, keccak256(leaf));
+  }
+
+  function claimed(bytes memory leaf) public view returns (bool) {
+    return rewardsClaimed[keccak256(leaf)];
+  }
+
+  function claimERC667(bytes memory leaf, address rewardProgramID, address rewardSafeOwner, bytes memory transferDetails, bool partialClaimAllowed) internal {
+      (
+        address payableToken,
+        uint256 amount
+      ) = abi.decode(transferDetails, (address, uint256));
+      // If the sender is willing to accept a partial claim and there isn't enough to cover the entire claim,
+      // then we can only claim the amount that is available _unless_ there is nothing left   
+      if (partialClaimAllowed && rewardBalance[rewardProgramID][payableToken] < amount && rewardBalance[rewardProgramID][payableToken] > 0) {
+        amount = rewardBalance[rewardProgramID][payableToken];
+      }
+      require(
+        IERC677(payableToken).balanceOf(address(this)) >= amount,
+        "Reward pool has insufficient balance"
+      );
+      require(
+        rewardBalance[rewardProgramID][payableToken] >= amount,
+        "Reward program has insufficient balance inside reward pool"
+      );
+
+      rewardsClaimed[keccak256(leaf)] = true;
+
+      rewardBalance[rewardProgramID][payableToken] = rewardBalance[
+        rewardProgramID
+      ][payableToken].sub(amount);
+      IERC677(payableToken).transfer(msg.sender, amount);
+
+      emit RewardeeClaim(
+        rewardProgramID,
+        rewardSafeOwner,
+        msg.sender,
+        payableToken,
+        amount
+      );
+  }
+
   function claim(
-    address rewardProgramID,
-    address payableToken,
-    uint256 amount,
-    bytes calldata proof
-  ) external returns (bool) {
-    require(amount > 0, "Cannot claim non-positive amount");
+    bytes calldata leaf,
+    bytes32[] calldata proof,
+    bool partialClaimAllowed
```

Comment:
do we really want to maintain `partialClaimAllowed`

Reply:
Probably a better name required here, this is for the requirement of "what if there's not enough in the pool for the full claim, do you still want to just get as much as you can".

Could maybe be two clear claim functions?